File: services/frontend/src/api/ota.ts
import { apiFetch } from "./client";

export type OtaCampaignRead = {
  id: string;
  customer_id: string;
  site_id: string | null;
  name: string;
  artifact_id: string | null;
  target_firmware_version: string | null;
  target_device_version_id: string | null;
  status: string;
  rollout_strategy: string | null;
  approval_status: string;
  created_by: string | null;
  approved_by: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
};

export type OtaCampaignTargetRead = {
  id: string;
  campaign_id: string;
  device_id: string;
  resolved_device_id: string | null;
  previous_device_version_id: string | null;
  target_device_version_id: string | null;
  current_firmware_version: string | null;
  target_firmware_version: string | null;
  status: string;
  progress_pct: number;
  failure_code: string | null;
  failure_message: string | null;
  last_status_at: string | null;
  completed_at: string | null;
  external_command_id: string | null;
  claimed_by: string | null;
  claimed_at: string | null;
  lease_expires_at: string | null;
  progress_phase: string | null;
  reported_ota_external_ref: string | null;
};

export type OtaCampaignDetailRead = OtaCampaignRead & {
  targets: OtaCampaignTargetRead[];
  /** Present for users with ota.launch when the campaign has a simulator poll token (after launch). */
  simulator_poll_url?: string | null;
  /** Same token as poll; POST terminal status without JWT. */
  simulator_status_url?: string | null;
};

export type OtaEventRead = {
  id: string;
  campaign_id: string;
  target_id: string | null;
  event_type: string;
  payload_json: Record<string, unknown> | null;
  created_at: string;
};

export type FirmwareArtifactRead = {
  id: string;
  customer_id: string;
  site_id: string | null;
  artifact_url: string;
  sha256: string;
  signature: string | null;
  signature_algorithm: string | null;
  size_bytes: number | null;
  release_notes: string | null;
  created_at: string;
};

export async function listOtaArtifacts(siteId: string) {
  return apiFetch<{ items: FirmwareArtifactRead[] }>(
    `/ota/artifacts?site_id=${encodeURIComponent(siteId)}`,
  );
}

export async function createOtaArtifact(body: {
  site_id: string;
  artifact_url: string;
  sha256: string;
  signature?: string | null;
  signature_algorithm?: string | null;
  size_bytes?: number | null;
  release_notes?: string | null;
}) {
  return apiFetch<FirmwareArtifactRead>("/ota/artifacts", { method: "POST", json: body });
}

export async function listOtaCampaigns(params?: { site_id?: string; status?: string }) {
  const sp = new URLSearchParams();
  if (params?.site_id) sp.set("site_id", params.site_id);
  if (params?.status?.trim()) sp.set("status", params.status.trim());
  const qs = sp.toString();
  return apiFetch<{ items: OtaCampaignRead[] }>(qs ? `/ota/campaigns?${qs}` : "/ota/campaigns");
}

export async function getOtaCampaign(campaignId: string) {
  return apiFetch<OtaCampaignDetailRead>(`/ota/campaigns/${encodeURIComponent(campaignId)}`);
}

export async function createOtaCampaign(body: {
  name: string;
  site_id: string;
  artifact_id?: string | null;
  target_firmware_version?: string | null;
  target_device_version_id?: string | null;
  rollout_strategy?: string | null;
}) {
  return apiFetch<OtaCampaignRead>("/ota/campaigns", { method: "POST", json: body });
}

export async function patchOtaCampaign(
  campaignId: string,
  body: {
    name?: string;
    artifact_id?: string | null;
    target_firmware_version?: string | null;
    rollout_strategy?: string | null;
    target_device_version_id?: string | null;
  },
) {
  return apiFetch<OtaCampaignRead>(`/ota/campaigns/${encodeURIComponent(campaignId)}`, {
    method: "PATCH",
    json: body,
  });
}

export async function addOtaCampaignTargets(campaignId: string, device_ids: string[]) {
  return apiFetch<{ added: OtaCampaignTargetRead[] }>(
    `/ota/campaigns/${encodeURIComponent(campaignId)}/targets`,
    { method: "POST", json: { device_ids } },
  );
}

export async function removeOtaCampaignTarget(campaignId: string, targetId: string) {
  await apiFetch(`/ota/campaigns/${encodeURIComponent(campaignId)}/targets/${encodeURIComponent(targetId)}`, {
    method: "DELETE",
  });
}

export async function submitOtaCampaign(campaignId: string) {
  return apiFetch<OtaCampaignRead>(`/ota/campaigns/${encodeURIComponent(campaignId)}/submit`, { method: "POST" });
}

export async function approveOtaCampaign(campaignId: string) {
  return apiFetch<OtaCampaignRead>(`/ota/campaigns/${encodeURIComponent(campaignId)}/approve`, { method: "POST" });
}

export async function launchOtaCampaign(campaignId: string) {
  return apiFetch<OtaCampaignRead>(`/ota/campaigns/${encodeURIComponent(campaignId)}/launch`, { method: "POST" });
}

export async function pauseOtaCampaign(campaignId: string) {
  return apiFetch<OtaCampaignRead>(`/ota/campaigns/${encodeURIComponent(campaignId)}/pause`, { method: "POST" });
}

export async function resumeOtaCampaign(campaignId: string) {
  return apiFetch<OtaCampaignRead>(`/ota/campaigns/${encodeURIComponent(campaignId)}/resume`, { method: "POST" });
}

export async function cancelOtaCampaign(campaignId: string) {
  return apiFetch<OtaCampaignRead>(`/ota/campaigns/${encodeURIComponent(campaignId)}/cancel`, { method: "POST" });
}

export async function listOtaCampaignEvents(campaignId: string, limit = 200) {
  return apiFetch<{ items: OtaEventRead[] }>(
    `/ota/campaigns/${encodeURIComponent(campaignId)}/events?limit=${encodeURIComponent(String(limit))}`,
  );
}

export async function reportOtaTargetStatus(
  body: {
    target_id: string;
    status: "success" | "failed" | "rolled_back" | "timeout" | "cancelled";
    command_id?: string | null;
    message?: string | null;
    ota_external_ref?: string | null;
  },
  opts?: {
    idempotencyKey: string;
    terminalOverride?: boolean;
  },
) {
  const idem = opts?.idempotencyKey?.trim();
  if (!idem) {
    throw new Error("reportOtaTargetStatus requires opts.idempotencyKey");
  }
  const headers: Record<string, string> = { "Idempotency-Key": idem };
  if (opts?.terminalOverride) headers["X-Ota-Terminal-Override"] = "true";
  return apiFetch<{
    target_id: string;
    campaign_id: string;
    device_id: string;
    status: string;
  }>("/ota/status", { method: "POST", json: body, headers });
}

File: services/api/app/schemas/ota.py
"""OTA status / campaign DTOs (Phases 4–5, 11)."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

TerminalOtaTargetStatus = Literal["success", "failed", "rolled_back", "timeout", "cancelled"]

OtaProgressPhase = Literal["acknowledged", "downloading", "verifying", "installing", "rebooting"]


class OtaTargetStatusReport(BaseModel):
    """POST /ota/status — report terminal (or tracked) state for one campaign target."""

    target_id: uuid.UUID
    status: TerminalOtaTargetStatus
    command_id: str | None = Field(None, max_length=255)
    message: str | None = Field(None, max_length=2000)
    ota_external_ref: str | None = Field(None, max_length=255)
    payload: dict[str, Any] | None = None


class OtaTargetStatusResponse(BaseModel):
    target_id: uuid.UUID
    campaign_id: uuid.UUID
    device_id: uuid.UUID
    status: str


class OtaCampaignCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    site_id: uuid.UUID
    artifact_id: uuid.UUID | None = None
    target_firmware_version: str | None = Field(None, max_length=128)
    target_device_version_id: uuid.UUID | None = None
    rollout_strategy: str | None = None


class OtaCampaignUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=255)
    artifact_id: uuid.UUID | None = None
    target_firmware_version: str | None = Field(None, max_length=128)
    rollout_strategy: str | None = None
    target_device_version_id: uuid.UUID | None = None


class OtaCampaignTargetRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    campaign_id: uuid.UUID
    device_id: uuid.UUID
    resolved_device_id: uuid.UUID | None = None
    previous_device_version_id: uuid.UUID | None = None
    target_device_version_id: uuid.UUID | None = None
    current_firmware_version: str | None = None
    target_firmware_version: str | None = None
    status: str
    progress_pct: int
    failure_code: str | None = None
    failure_message: str | None = None
    last_status_at: datetime | None = None
    completed_at: datetime | None = None
    external_command_id: str | None = None
    claimed_by: str | None = None
    claimed_at: datetime | None = None
    lease_expires_at: datetime | None = None
    progress_phase: str | None = None
    reported_ota_external_ref: str | None = None


class OtaCampaignRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    customer_id: uuid.UUID
    site_id: uuid.UUID | None = None
    name: str
    artifact_id: uuid.UUID | None = None
    target_firmware_version: str | None = None
    target_device_version_id: uuid.UUID | None = None
    status: str
    rollout_strategy: str | None = None
    approval_status: str
    created_by: uuid.UUID | None = None
    approved_by: uuid.UUID | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime


class OtaCampaignDetailRead(OtaCampaignRead):
    targets: list[OtaCampaignTargetRead] = Field(default_factory=list)
    simulator_poll_url: str | None = Field(
        default=None,
        description="Public HTTPS poll URL including secret token; only returned for users with ota.launch.",
    )
    simulator_status_url: str | None = Field(
        default=None,
        description="Public HTTPS POST URL (same token as poll) for terminal status; only for users with ota.launch.",
    )


class OtaCampaignListResponse(BaseModel):
    items: list[OtaCampaignRead]


class OtaAddTargetsRequest(BaseModel):
    device_ids: list[uuid.UUID] = Field(..., min_length=1)


class OtaAddTargetsResponse(BaseModel):
    added: list[OtaCampaignTargetRead]


class OtaEventRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    campaign_id: uuid.UUID
    target_id: uuid.UUID | None = None
    event_type: str
    payload_json: dict[str, Any] | None = None
    created_at: datetime


class OtaEventListResponse(BaseModel):
    items: list[OtaEventRead]


class FirmwareArtifactCreate(BaseModel):
    site_id: uuid.UUID
    artifact_url: str = Field(..., min_length=1, max_length=8000)
    sha256: str = Field(..., min_length=1, max_length=128)
    signature: str | None = Field(None, max_length=16000)
    signature_algorithm: str | None = Field(None, max_length=64)
    size_bytes: int | None = Field(None, ge=0)
    release_notes: str | None = Field(None, max_length=32000)


class FirmwareArtifactRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    customer_id: uuid.UUID
    site_id: uuid.UUID | None = None
    artifact_url: str
    sha256: str
    signature: str | None = None
    signature_algorithm: str | None = None
    size_bytes: int | None = None
    release_notes: str | None = None
    created_at: datetime


class FirmwareArtifactListResponse(BaseModel):
    items: list[FirmwareArtifactRead]


class OtaExecutorArtifactBlock(BaseModel):
    url: str | None = None
    sha256: str | None = None
    signature: str | None = None
    signature_algorithm: str | None = None
    size_bytes: int | None = None
    release_notes: str | None = None


class OtaExecutorWorkItemRead(BaseModel):
    campaign_id: uuid.UUID
    target_id: uuid.UUID
    device_id: uuid.UUID
    device_display_id: str
    resolved_device_id: uuid.UUID | None = None
    target_firmware_version: str | None = None
    target_device_version_id: uuid.UUID | None = None
    artifact: OtaExecutorArtifactBlock = Field(default_factory=OtaExecutorArtifactBlock)


class OtaExecutorWorkListResponse(BaseModel):
    items: list[OtaExecutorWorkItemRead]
    next_cursor: str | None = None


class OtaSimulatorPublicPollResponse(BaseModel):
    """Public poll payload for upstream / lab OTA simulators (token-authenticated, no JWT)."""

    campaign_id: uuid.UUID
    campaign_name: str
    campaign_status: str
    hint: str = Field(
        ...,
        description="ok when items may be non-empty; campaign_not_releasing_work when not running",
    )
    items: list[OtaExecutorWorkItemRead] = Field(default_factory=list)
    next_cursor: str | None = None


class OtaTargetClaimRequest(BaseModel):
    executor_id: str = Field(..., min_length=1, max_length=255)
    lease_seconds: int = Field(300, ge=30, le=86400)


class OtaProgressReport(BaseModel):
    target_id: uuid.UUID
    phase: OtaProgressPhase
    message: str | None = Field(None, max_length=2000)
    payload: dict[str, Any] | None = None
